### PR TITLE
#444 - overtime application process

### PR DESCRIPTION
**Pull request by kamilpiech97**
Connected with #444.

Added the ability to add an overtime application, where the employee can add the date and time `from` and `to`, `type` of preferred overtime settlement. Added tests.

Preview:

[Screencast from 10-06-24 13:47:25.webm](https://github.com/blumilksoftware/toby/assets/29115024/cc631290-154b-4242-9671-98709ed5322d)

### CODE REVIEW

File: app/Actions/OvertimeRequest/AcceptAsTechnicalAction.php [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+use Toby\Domain\OvertimeRequestStateManager;
+use Toby\Domain\VacationTypeConfigRetriever;
+use Toby\Models\OvertimeRequest;
+use Toby\Models\User;
+
+class AcceptAsTechnicalAction
+{
+    public function __construct(
+        protected OvertimeRequestStateManager $stateManager,
+        protected VacationTypeConfigRetriever $configRetriever,
```
Comment (krzysztofrewak) [inline app/Actions/OvertimeRequest/AcceptAsTechnicalAction.php:16]: Is it used anywhere?

File: app/Domain/OvertimeCalculator.php [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+namespace Toby\Domain;
+
+use Carbon\Carbon;
+use Carbon\CarbonInterface;
+
+class OvertimeCalculator
+{
+    public function calculateHours(CarbonInterface $from, CarbonInterface $to): int
+    {
+        $hours = Carbon::create($from)->diffInMinutes($to, true);
```
Comment (krzysztofrewak) [inline app/Domain/OvertimeCalculator.php:14]: Please use named argument in `true`.

File: app/Enums/SettlementType.php [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    public static function casesToSelect(): array
+    {
+        $cases = collect(SettlementType::cases());
+
+        return $cases->map(
+            fn(SettlementType $enum): array => [
+                "label" => $enum->label(),
+                "value" => $enum->value,
+            ],
+        )->toArray();
```
Comment (krzysztofrewak) [inline app/Enums/SettlementType.php:26]: Isn't that overkill?

```suggestion
        return array_map(
            fn(SettlementType $enum): array => [
                "label" => $enum->label(),
                "value" => $enum->value,
            ],
            SettlementType::cases(),
        )
```

File: app/Http/Controllers/OvertimeRequestController.php [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+            ->overtimeRequests()
+            ->whereBelongsTo($yearPeriodRetriever->selected())
+            ->states(OvertimeRequestStatesRetriever::failedStates())
+            ->count();
+
+        $settled = $request->user()
+            ->overtimeRequests()
+            ->whereBelongsTo($yearPeriodRetriever->selected())
+            ->states(OvertimeRequestStatesRetriever::settledStates())
+            ->count();
```
Comment (krzysztofrewak) [inline app/Http/Controllers/OvertimeRequestController.php:65]: Maybe we could cache it somehow?

File: app/Http/Controllers/OvertimeRequestController.php [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+            "filters" => [
+                "status" => $status,
+            ],
+        ]);
+    }
+
+    public function indexForApprovers(
+        Request $request,
+        YearPeriodRetriever $yearPeriodRetriever,
+    ): RedirectResponse|Response {
```
Comment (krzysztofrewak) [inline app/Http/Controllers/OvertimeRequestController.php:85]: Maybe throwing an exception with redirect would be better?

File: app/Http/Resources/OvertimeRequestResource.php [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    public function toArray($request): array
+    {
+        $user = $request->user();
+
+        return [
+            "id" => $this->id,
+            "name" => $this->name,
+            "user" => new SimpleUserResource($this->user),
+            "state" => $this->state,
+            "from" => $this->from->format("Y-m-d H:i"),
```
Comment (krzysztofrewak) [inline app/Http/Resources/OvertimeRequestResource.php:26]: Maybe you could move this very specific format (`""Y-m-d H:i"`) to some helper/constant?

**Review by krzysztofrewak — CHANGES_REQUESTED** [review]

File: app/Actions/OvertimeRequest/AcceptAsTechnicalAction.php (hunk 1 above)
Comment (kamilpiech97) [reply app/Actions/OvertimeRequest/AcceptAsTechnicalAction.php]: removed

File: app/Http/Controllers/OvertimeRequestController.php (hunk 4 above)
Comment (kamilpiech97) [reply app/Http/Controllers/OvertimeRequestController.php]: added query cache, for vacation requests too

File: app/Http/Controllers/OvertimeRequestController.php (hunk 5 above)
Comment (kamilpiech97) [reply app/Http/Controllers/OvertimeRequestController.php:89]: added `abort(403)`

File: app/Http/Resources/OvertimeRequestResource.php (hunk 6 above)
Comment (kamilpiech97) [reply app/Http/Resources/OvertimeRequestResource.php]: added Helper

**Review by EwelinaSkrzypacz — COMMENTED** [review]
1. When I create an overtime request, the number of hours is always 0 (when I use localhost address, because when I'm using `http://toby.blumilk.localhost/overtime/requests/create` it is fine

image

2. Please fix this validation massage, because it's a bit unclear.

image

3. For me default time it's shouldn't be midnight - I think that most of overtime is afternoon, so maybe it should be like 16:00? Or 15:00? 

4. As an administrator, I don't think we need "Moje nadgodziny" - maybe just do like we have requests? I mean we have just "Requests". 

image

5. Date format - in all appliction we have dd.mm.yyy

image

6. As an employee should I be able to add overtime request, when I have pending/accepted vacation?

7. I think we should add here counter, when user has some overtime to accept

image

8. "Rozliczenie nadgodzin oznaczy nadgodziny jako rozliczone." - can we write this somehow different? 

9. I think that admin should have possibility to cancel settled overtime; just in case - like we have canceling approved vacation requests.

10. Now numeration for overtime and requests looks the same. Maybe we should distinguish it somehow?

11. As I understand accepting/declining overtime in Slack will be implemented in another pr? If so, please create issue for that or at least add comment in issue - I don't want to forget about it.

12. For now, as an administrative approver I can accept overtime. I thought that only technical approver/admin can do that.

**Comment by kamilpiech97** [comment]
> 1. When I create an overtime request, the number of hours is always 0 (when I use localhost address, because when I'm using `http://toby.blumilk.localhost/overtime/requests/create` it is fine
> 
> image
> 
> 2. Please fix this validation massage, because it's a bit unclear.
> 
> image
> 
> 3. For me default time it's shouldn't be midnight - I think that most of overtime is afternoon, so maybe it should be like 16:00? Or 15:00?
> 4. As an administrator, I don't think we need "Moje nadgodziny" - maybe just do like we have requests? I mean we have just "Requests".
> 
> image
> 
> 5. Date format - in all appliction we have dd.mm.yyy
> 
> image
> 
> 6. As an employee should I be able to add overtime request, when I have pending/accepted vacation?
> 7. I think we should add here counter, when user has some overtime to accept
> 
> image
> 
> 8. "Rozliczenie nadgodzin oznaczy nadgodziny jako rozliczone." - can we write this somehow different?
> 9. I think that admin should have possibility to cancel settled overtime; just in case - like we have canceling approved vacation requests.
> 10. Now numeration for overtime and requests looks the same. Maybe we should distinguish it somehow?
> 11. As I understand accepting/declining overtime in Slack will be implemented in another pr? If so, please create issue for that or at least add comment in issue - I don't want to forget about it.
> 12. For now, as an administrative approver I can accept overtime. I thought that only technical approver/admin can do that.

1. Lets create issue for that.
2. fixed
3. I will stay with that
4. Fixed
5. Fixed
6. Let's ask Kamila but AFAIK we do not want create compliated valdation for overtime process
7. Added
8. Fixed
9. Added
10. Changed
11. In next PR
12. I will stay with that

**Review by EwelinaSkrzypacz — COMMENTED** [review]
When I logged as an employee and click "Moje nadgodziny" I see error:

image

**Comment by kamilpiech97** [comment]
> When I logged as an employee and click "Moje nadgodziny" I see error:
> 
> image

Run seeder?

**Review by EwelinaSkrzypacz — COMMENTED** [review]
1. As a employee I can settle overtime. Is it okay for us?
2. When I click "Moje nadgodziny" in menu and click button "Dodaj wniosek" (so I'm on the page with creating overtime) on the menu the text "Moje nadgodziny" is not highlighted - in requests it is.
3. So, when I have settled overtime for example 19.06.2024 15:00 - 19.06.2024 17:00, I can create another overtime 19.06.2024 16:00 - 19.06.2024 18:00. Is this okay for us?

**Comment by kamilpiech97** [comment]
> 1. As a employee I can settle overtime. Is it okay for us?
> 2. When I click "Moje nadgodziny" in menu and click button "Dodaj wniosek" (so I'm on the page with creating overtime) on the menu the text "Moje nadgodziny" is not highlighted - in requests it is.
> 3. So, when I have settled overtime for example 19.06.2024 15:00 - 19.06.2024 17:00, I can create another overtime 19.06.2024 16:00 - 19.06.2024 18:00. Is this okay for us?

1. yep
2. fixed
3. fixed, added edge cases

**Review by EwelinaSkrzypacz — DISMISSED** [review]

**Review by EwelinaSkrzypacz — DISMISSED** [review]

File: app/Http/Controllers/OvertimeRequestController.php [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+            ->whereBelongsTo($yearPeriodRetriever->selected())
+            ->states(OvertimeRequestStatesRetriever::successStates())
+            ->cache(key: "overtimeStats")
+            ->count();
+
+        $failed = $request->user()
+            ->overtimeRequests()
+            ->whereBelongsTo($yearPeriodRetriever->selected())
+            ->states(OvertimeRequestStatesRetriever::failedStates())
+            ->cache(key: "overtimeStats")
```
Comment (krzysztofrewak) [inline app/Http/Controllers/OvertimeRequestController.php:64]: Does it cache it globally or per user?

Comment (kamilpiech97) [reply app/Http/Controllers/OvertimeRequestController.php:64]: globally but only for 60 seconds

Comment (kamilpiech97) [reply app/Http/Controllers/OvertimeRequestController.php:64]: it is for index page where u switching between tabs with specific states

**Review by krzysztofrewak — DISMISSED** [review]

**Review by EwelinaSkrzypacz — APPROVED** [review]

**Pull request merged**